giraffe/src/utils/geo.ts

```diff
@@ -0,0 +1,205 @@
+  width: number,
+  height: number,
+  lon: number,
+  lat: number,
+  zoom: number
+): GeoQueryVariables => {
+  const pixelRadius = Math.sqrt(width * width + height * height) / 2
+  const metersPerPixel =
+    (circumEarth * Math.abs(Math.cos((lat * Math.PI) / 180))) /
+    Math.pow(2, zoom + 8)
+  return {lon, lat, radius: (pixelRadius * metersPerPixel) / 1000}
+}
+
+export const formatCircleMarkerRowInfo = (
+  properties: GeoCircleViewLayer,
+  table: GeoTable,
+  index
+) => {
+  const result = []
+  const timeValue = table.getTimeString(index)
+  if (timeValue) {
+    result.push({
+      key: '_time',
+      name: 'Time',
+      type: 'string',
+      values: [timeValue],
+    })
+  }
+  const {colorField, radiusField} = properties
+  const radiusValue = table.getValue(index, radiusField)
+  const {radiusDimension} = properties
+  const radiusInfo = formatValue(
+    radiusField,
+    'Radius',
+    radiusValue,
+    radiusDimension
+  )
+  if (radiusInfo) {
+    result.push(radiusInfo)
+  }
+  const colorValue = table.getValue(index, colorField)
+  const {colorDimension} = properties
+  const colorInfo = formatValue(colorField, 'Color', colorValue, colorDimension)
+  if (colorInfo) {
+    result.push(colorInfo)
+  }
+  return result
+}
+
+export const formatPointLayerRowInfo = (
+  properties: GeoPointMapViewLayer,
+  table: GeoTable,
+  index
+) => {
+  const result = []
+  const timeValue = table.getTimeString(index)
+  if (timeValue) {
+    result.push({
+      key: '_time',
+      name: 'Time',
+      type: 'string',
+      values: [timeValue],
+    })
+  }
+  const colorValue = table.getValue(index, properties.colorField)
+  const {colorDimension} = properties
+  const colorInfo = formatValue(
+    properties.colorField,
+    'Color',
+    colorValue,
+    colorDimension
+  )
+  if (colorInfo) {
+    result.push(colorInfo)
+  }
+  return result
+}
+
+const getClusterValues = (cluster, result: number[]): GeoPointMapViewLayer => {
+  let clusterRenderingProperties
+  for (const marker of cluster._markers) {
+    result.push(marker.value)
+    clusterRenderingProperties = marker.clusterRenderingProperties
+  }
+  for (const childCluster of cluster._childClusters) {
+    const properties = getClusterValues(childCluster, result)
+    if (properties) {
+      clusterRenderingProperties = properties
+    }
+  }
+  return clusterRenderingProperties
+}
+
+const getClusterValue = (
+  cluster
+): {value: number; properties: GeoPointMapViewLayer} => {
+  const values: number[] = []
+  const properties = getClusterValues(cluster, values)
+  const aggregation =
+    properties.clusterAggregationFunction || ClusterAggregation.max
+  switch (aggregation) {
+    case ClusterAggregation.max:
+      return {value: Math.max(...values), properties}
+    case ClusterAggregation.min:
+      return {value: Math.min(...values), properties}
+    case ClusterAggregation.mean:
+      return {
+        value: values.reduce((a, x) => a + x, 0) / values.length,
```

Comment:
if `values` is empty, this crashes because of divide by zero, so we should check for it being zero and handle it.